R/utils.R

```diff
@@ -54,3 +54,21 @@ deparse_to_one_line <- function(x) {
          paste(deparse(x), collapse = ' '),
          perl = TRUE)
 }
+
+#' Catch the log header
+#' @return string
+#' @keywords internal
+#' @param level see \code{\link{log_levels}}
+#' @param namespace string
+catch_base_log <- function(level, namespace) {
+    orginal_appender <- log_appender()
+    log_appender(appender_console)
+    # catch error, warning or message
+    res <- capture.output(log_level(level = level, namespace = namespace), type = "message")
+    if (length(res) == 0) {
+        # catch output
+        res <- capture.output(log_level(level = level, namespace = namespace), type = "output")
+    }
```

Comment:
do we need to check both `stderr` and `stdout`? setting `appender_console` (which is an alias to `appender_stderr`) should (in theory) write to the former, and nothing to the latter.

Reply:
I added the `log_appender(appender_console)` call as I got double prints in some tests. Preciously for helpers testing where there is used `log_appender(appender_stdout)`.  `catch_base_log` should be neutral for the rest of the session, that is why I am doing stderr redirection (2>).

Reply:
hmmm maybe sth like:
```
orginal_appender <- log_appender()
log_appender(logger::appender_stderr)
# catch error, warning or message
res <- capture.output(log_level(level = level, namespace = namespace), type = "message")
log_appender(logger::appender_stdout)
res <- capture.output(log_level(level = level, namespace = namespace), type = "output")
log_appender(orginal_appender)
res
```

Reply:
double prints should not happen, except if namespaces and/or indexes are messed up. I think passing `namespace` in the other related comment, and systematically everywhere in this function might resolve it?

Reply:
I think this is ok as log_appender is additional print and not influence the base one.
```
> logger::log_appender(logger::appender_stderr)
> logger::log_error("aa")
ERROR [2021-12-02 09:28:17] aa
> logger::log_info("aa")
INFO [2021-12-02 09:28:20] aa
```

Reply:
You are right namespaces were a clue. More than that I have to use fallback_namespace.

```r
catch_base_log <- function(level, namespace) {
    namespace <- fallback_namespace(namespace)
    orginal_appender <- log_appender(namespace = namespace)
    log_appender(appender_console, namespace = namespace)
    # catch error, warning or message
    res <- capture.output(log_level(level = level, namespace = namespace), type = "message")
    log_appender(orginal_appender, namespace = namespace)
    res
}
```

1. getting namespace and if not exists then return global.
2. getting current namespace appender, to be able to return to it later.
3. temporally setting appender_console (stderr) inside the namespace and catch (run) capture.output(log_layout(level, namespace = namespace), type = "message").
4. Go back to original appender and apply it to the namespace
5. return catched header (from log_layout)

We get a log_layout header and do not influence the environment

Reply:
Please read on github as I added small edit.

Reply:
In the last commit I added testthat tests for the `catch_base_log` function.

Reply:
I start to consider `.topcall` so the function is even more secure when the layout depends on where the function is called. I added `.topcall` argument to log_separator too. More tests are added for this scenerios.